freqtrade/optimize/backtesting.py

```diff
@@ -365,35 +382,42 @@ def _get_sell_trade_entry(
             return bt_res
         return None
 
-    def backtest(self, args: Dict) -> DataFrame:
+    def backtest(self, processed, stake_amount: float,
+                 start_date, end_date,
+                 max_open_trades: int = 0, position_stacking: bool = False,
+                 populate_indicators: bool = False,
+                 populate_buy: bool = True,
+                 populate_sell: bool = True) -> DataFrame:
         """
-        Implements backtesting functionality
+        Implement backtesting functionality
 
         NOTE: This method is used by Hyperopt at each iteration. Please keep it optimized.
         Of course try to not have ugly code. By some accessor are sometime slower than functions.
-        Avoid, logging on this method
-
-        :param args: a dict containing:
-            stake_amount: btc amount to use for each trade
-            processed: a processed dictionary with format {pair, data}
-            max_open_trades: maximum number of concurrent trades (default: 0, disabled)
-            position_stacking: do we allow position stacking? (default: False)
-        :return: DataFrame
+        Avoid extensive logging in this method and functions it calls.
+
+        :param processed: a processed dictionary with format {pair, data}
+        :param stake_amount: amount to use for each trade
+        :param start_date: backtesting timerange start datetime
+        :param end_date: backtesting timerange end datetime
+        :param max_open_trades: maximum number of concurrent trades, <= 0 means unlimited
+        :param position_stacking: do we allow position stacking?
+        :return: DataFrame with trades (results of backtesting)
         """
-        # Arguments are long and noisy, so this is commented out.
-        # Uncomment if you need to debug the backtest() method.
-#        logger.debug(f"Start backtest, args: {args}")
-        processed = args['processed']
-        stake_amount = args['stake_amount']
-        max_open_trades = args.get('max_open_trades', 0)
-        position_stacking = args.get('position_stacking', False)
-        start_date = args['start_date']
-        end_date = args['end_date']
+        logger.debug(
+            f"Run backtest, stake_amount: {stake_amount}, "
+            f"start_date: {start_date}, end_date: {end_date}, "
+            f"max_open_trades: {max_open_trades}, position_stacking: {position_stacking}"
+        )
```

Comment:
In my measurements, it's not worth, f-strings are fast enough:
```
$ python -m timeit -s "import logging; logger=logging.getLogger(); logger.setLevel(logging.INFO); from datetime import datetime; stake_amount=0.05; start_date=datetime.utcnow(); end_date=datetime.now(); max_open_trades=10; position_stacking=False" 'logger.debug("Run backtest, stake_amount: %s, start_date: %s, end_date: %s, max_open_trades: %s, position_stacking: %s" % (stake_amount, start_date, end_date, max_open_trades, position_stacking))'
100000 loops, best of 3: 3.52 usec per loop
```

```
$ python -m timeit -s "import logging; logger=logging.getLogger(); logger.setLevel(logging.INFO); from datetime import datetime; stake_amount=0.05; start_date=datetime.utcnow(); end_date=datetime.now(); max_open_trades=10; position_stacking=False" 'logger.debug(f"Run backtest, stake_amount: {stake_amount}, " f"start_date: {start_date}, end_date: {end_date}, " f"max_open_trades: {max_open_trades}, position_stacking: {position_stacking}")'
100000 loops, best of 3: 3.96 usec per loop
```

(python 3.6.9)

Reply:
And without logging, compare creation of strings (for completeness of the measurements, the above assumes lazyness of '%' evaluations of attributes of the logging arguments):

```
$ python -m timeit -s "from datetime import datetime; stake_amount=0.05; start_date=datetime.utcnow(); end_date=datetime.now(); max_open_trades=10; position_stacking=False" 's = "Run backtest, stake_amount: %s, start_date: %s, end_date: %s, max_open_trades: %s, position_stacking: %s" % (stake_amount, start_date, end_date, max_open_trades, position_stacking)'
100000 loops, best of 3: 2.94 usec per loop
```
```
$ python -m timeit -s "from datetime import datetime; stake_amount=0.05; start_date=datetime.utcnow(); end_date=datetime.now(); max_open_trades=10; position_stacking=False" 's = f"Run backtest, stake_amount: {stake_amount}, " f"start_date: {start_date}, end_date: {end_date}, " f"max_open_trades: {max_open_trades}, position_stacking: {position_stacking}"'
100000 loops, best of 3: 3.28 usec per loop
```

Reply:
The above was wrong, this variant employs lazy evaluations (no '%' sign between formatting string and the parameters
```
$ python -m timeit -s "import logging; logger=logging.getLogger(); logger.setLevel(logging.INFO); from datetime import datetime; stake_amount=0.05; start_date=datetime.utcnow(); end_date=datetime.now(); max_open_trades=10; position_stacking=False" 'logger.debug(f"Run backtest, stake_amount: {stake_amount}, " f"start_date: {start_date}, end_date: {end_date}, " f"max_open_trades: {max_open_trades}, position_stacking: {position_stacking}")'
100000 loops, best of 3: 4.03 usec per loop
```
```
$ python -m timeit -s "import logging; logger=logging.getLogger(); logger.setLevel(logging.INFO); from datetime import datetime; stake_amount=0.05; start_date=datetime.utcnow(); end_date=datetime.now(); max_open_trades=10; position_stacking=False" 'logger.debug("Run backtest, stake_amount: %s, start_date: %s, end_date: %s, max_open_trades: %s, position_stacking: %s", stake_amount, start_date, end_date, max_open_trades, position_stacking)'
1000000 loops, best of 3: 0.629 usec per loop
```
-- '%' formatting is 5-7 times faster

Reply:
But this is printed (evaluated and INFO level) once per epoch, so it will only give less for 1 sec for 1000 epochs (~4 microsec per epoch)... Let's leave it as it is?

Reply:
you say info level ... but the code says debug . . . ? 

Apart from that, i trust your judgement that if you say it'll be ok for hyperopt - then it's ok (it's not really relevant for backtesting since it's only ever called once (per strategy) ... just wanted to point it out ... since this is something we can easily overlook.

Reply:
My conclusion: f-strings are okay in the modern python (it's an active part of the language and is constantly improved/optimized), changing them in the logging into old-style lazy '%' formatting only worth when the parameters to be logged are complicated for evaluation. Simple logging of the values that are ready in the vars do not worth it...

Reply:
> you say info level ... but the code says debug . . . ?

The strings are printed into the log stream with the log level set to debug and only evaluated with the log level set to info. We do not worry about first (since printing takes most time) and only care about the latter -- when no debug messages are printed, but the strings are evaluated, if they are not lazy...